direct parsing fuzzing test

tyler92 opened the pull request:
Two issues were found by the new fuzzing target: https://github.com/boostorg/json/issues/1047 and https://github.com/boostorg/json/issues/1048

I think detecting memory issues in case of new changes or refactoring might be useful.
I'm not sure about code formatting for the project, I will appreciate your feedback.

grisumbras commented:
While the idea is interesting, I am sceptical fuzzing with random strings can be effective when the parser expects a very specific structure of the input. It will probably always fail on the second (`on_object_begin`), the third (`on_key`), or the fourth (`on_bool`) event.

tyler92 commented:
According to coverage reports, it works well enough to find valid inputs, these two issues were found quite quickly. We can provide a dictionary for the fuzzer to help it a bit if we want to improve the performance.

sdarwin commented:
It may be a coincidence. This pull request mentions "detecting memory issues". The drone agent on IBM s390 crashed with out-of-memory errors.  I have rebooted the server.

grisumbras commented:
It's also possible that linking the new fuzzer requires too much memory (variant parsers do do that).

tyler92 commented:
> I am sceptical fuzzing with random strings can be effective when the parser expects a very specific structure of the input. It will probably always fail on the second (on_object_begin), the third (on_key), or the fourth (on_bool) event.

Just for an experiment: I reverted the fix for https://github.com/boostorg/json/issues/1047 and measured the time required to find that bug. Without a dictionary and with an empty corpus (the worst scenario) it takes about 10 seconds. The reason is the fuzzer doesn't have to find a valid input with all required fields. It's enough to generate a JSON with some fields to verify parser for a specific type only.

Adding corpus doesn't improve speed much. But I've added two valid JSONs to seed files, and it gives the fuzzer the ability to reach the "success" point. With this change, the fuzzer finds https://github.com/boostorg/json/issues/1047 in less than one second. Also, I changed C++ version to C++14 to avoid potential OOM.

Please let me know if it makes sense

grisumbras commented on fuzzing/fuzz_direct_parse.cpp:
```diff
@@ -0,0 +1,92 @@
+    std::tuple<std::vector<std::string>, std::vector<double>> t3;
+
+#ifndef BOOST_NO_CXX17_HDR_VARIANT
+    std::variant<bool, std::uint64_t, std::int64_t, double, std::string> v;
```
Please, replace with `boost::variant2::variant`. It will reduce the amount of macro switching necessary.

tyler92 replied:
Done

grisumbras commented on .github/workflows/run_fuzzer.yml:
```diff
@@ -60,6 +60,7 @@ jobs:
         buildtype: 'boost'
         path: 'head'
         toolset: clang-18
+        cxxstd: 14
```
Why have you reduced this from 17 to 14?

tyler92 replied:
I wanted to make sure this MR has a chance to be approved and avoid potential OOMs on the drone agent before (although I'm not sure if it's related to the current MR). I reverted to 17; please let me know if there are any issues with it.

grisumbras reviewed (approved):
Everything looks good. Can you please squsah and rebase on the current develop (if it's not already). I can also do it manually myself, but then the PR won't be considered merged by GitHub.

tyler92 commented:
> Can you please squsah and rebase on the current develop (if it's not already). I can also do it manually myself, but then the PR won't be considered merged by GitHub.

Done

grisumbras commented:
Thank you for your contribution.
